yarl/__init__.py

```diff
@@ -557,16 +557,26 @@ def with_port(self, port):
                                                                port)),
                    encoded=True)
 
-    def with_query(self, query):
+    def with_query(self, *args, **kwargs):
         """Return a new URL with query part replaced.
 
         Accepts any Mapping (e.g. dict, multidict.MultiDict instances)
         or str, autoencode the argument if needed.
 
+        It also can take an arbitrary number of keyword arguments.
+
         Clear query if None is passed.
 
         """
         # N.B. doesn't cleanup query/fragment
+
+        if kwargs:
+            query = kwargs
+        elif len(args) == 1:
+            query = args[0]
+        else:
+            raise TypeError("Invalid query type")
```

Comment:
It's `ValueError` actually.
Also please improve the message -- it should have a clue for proper format.